tracing-subscriber/src/layer.rs

```diff
@@ -407,6 +420,63 @@ pub struct Layered<L, I, S = I> {
     _s: PhantomData<fn(S)>,
 }
 
+/// A [`Subscriber`] composed of a `Subscriber` wrapped by one or more
+/// [`Layer`]s and [`Filter`]s.
+///
+/// [`Layer`]: ../layer/trait.Layer.html
+/// [`Subscriber`]: https://docs.rs/tracing-core/latest/tracing_core/trait.Subscriber.html
+/// [`Filter`]: ../layer/trait.Layer.html
+#[derive(Clone, Debug)]
+pub struct Filtered<L, S, F> {
+    layer: L,
+    filter: F,
+    _s: PhantomData<fn(S)>,
+}
+
+/// `Filter` enables the filtering of composed [`Layers`].
+///
+/// [`Layer`]: ../layer/trait.Layer.html
+pub trait Filter<L, S>
+where
+    L: Layer<S>,
+    S: Subscriber + 'static,
+{
+    /// Filter on a specific span or event's metadata.
+    fn filter(&self, metadata: &Metadata<'_>, ctx: &Context<'_, S>) -> bool;
+}
+
+impl<L, S, F> Filter<L, S> for F
+where
+    L: Layer<S>,
+    S: Subscriber + 'static,
+    F: Fn(&Metadata<'_>, &Context<'_, S>) -> bool,
+{
+    fn filter(&self, metadata: &Metadata<'_>, ctx: &Context<'_, S>) -> bool {
+        (self)(metadata, ctx)
+    }
+}
+
+impl<L, S, F> Layer<S> for Filtered<L, S, F>
+where
+    L: Layer<S>,
+    S: Subscriber + 'static,
+    F: Filter<L, S> + 'static,
+{
```

Comment:
I think this needs to implement the `enabled` and `register_callsite` layer fns. I think we will also want to change the rules for how layers handle filters — I'd really like to get it so that if all the per-layer filters disable something, the call to `enabled` will return `false`.

Reply:
Yeah, I think this would require thinking a little about interest evaluation orders, which I have _not_ done.